### PR TITLE
Link model updates

### LINKED ISSUE #30: Link model FK admin
It was mentioned in #21 that it might be better to make Link and FileLink abstract models and have the models based on those define a FK to film. This is fine except thus far it's proven very problematic to display in the admin site on the Film form in a meaningful way. They need to be dispersed throughout the form and be under certain headings, but the behaviour of inlines is to be included after the form.
I have tried referencing them as functions to include in fields, tried the `fieldsets_with_inlines` package and have tried creating inline formsets to include on FilmAdminForm, all to no avail. 
At present this seems more hassle than necessary since the form works perfectly fine without these changes but as the admin view is the last thing to be completed I am still happy to figure out how to display them properly.
The code can be found at https://github.com/UCL-ARC/museum_of_dreams/tree/link-model-updates

### PR DESCRIPTION
**Pull request by acholyn**
Updating link models and those based thereon to have a FK to film. This includes migration and admin changes/updates.

closes #30 and #26 though the latter may not be fixed on AWS
addresses part of #14 by customising the home dashboard - if it's possible to prevent access to `/admin/mod_app` and just use `/admin` that might be better

It also contains resolutions to some things raised in #29 , which this should be merged into

### CODE REVIEW

**Comment by tcouch** [comment]
If I run `./manage.py migrate` from a clean start I get `ValueError: Related model 'mod_app.filelink' cannot be resolved`. Putting all the `migrations.RemoveField(` lines before the migrations.DeleteModel(` lines in 0001b seems to fix it.

File: mod_app/admin/film_admin.py [hunk 1]
```diff
@@ -28,11 +43,36 @@ class TRInline(admin.TabularInline):
...
-    inlines = [FilmAnalysisInline, TRInline]
+    class Media:
+        css = {
+            "all": ("admin/css/custom.css",),
+        }
+
+    autocomplete_fields = ["genre"]
+    search_fields = ["title", "alt_titles"]
+    formfield_overrides = {
+        models.TextField: {"widget": CKEditorWidget},
```
Comment (tcouch) [inline mod_app/admin/film_admin.py:54]: If it's not too much effort, I'd suggest paring down the toolbar on some of the film admin text fields like cast, credits (I doubt they'll need emojis 😸)

File: mod_app/admin/film_admin.py [hunk 2]
```diff
@@ -103,32 +126,31 @@ def safe_comments(self, obj):
...
-                    ("programmes", "pub_mat"),
-                ),
-            },
+            {"fields": []},
         ),
+        (None, {"classes": ("placeholder scripts-group",), "fields": ()}),
+        (None, {"classes": ("placeholder pressbooks-group",), "fields": ()}),
+        (None, {"classes": ("placeholder programmes-group",), "fields": ()}),
+        (None, {"classes": ("placeholder publicitys-group",), "fields": ()}),
         (
```
Comment (tcouch) [inline mod_app/admin/film_admin.py:137]: Is the idea here that the empty grey printed materials section acts as a sort of section header/divider? I found it a little confusing when trying to understand what I was looking at, but I presume you'd have to get into writing custom templates or something to get exactly what you're after?

File: mod_app/migrations/0001b_remove_link_models.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+class Migration(migrations.Migration):
+    dependencies = [
+        ("mod_app", "0001_initial"),
+    ]
+
+    operations = [
+        migrations.DeleteModel(name="Link"),
+        migrations.DeleteModel(name="Drawing"),
+        migrations.DeleteModel(name="Postcard"),
```
Comment (tcouch) [inline mod_app/migrations/0001b_remove_link_models.py:14]: See my comment about RemoveField having to be before DeleteModel

File: mod_app/models/film_model.py [hunk 4]
```diff
@@ -1,21 +1,27 @@
...
     Publicity,
-    Script,
-    Source,
     Still,
-    Tag,
+    Postcard,
 )
 
 
+def validate_format_other(value, format_type):
```
Comment (tcouch) [inline mod_app/models/film_model.py:20]: This doesn't seem to work?

File: mod_app/models/film_model.py [hunk 5]
```diff
@@ -87,10 +85,17 @@ def __str__(self):
...
+        verbose_name="format",
+        choices=FORMAT_CHOICES,
     )
     format_other = models.CharField(
-        max_length=255, blank=True, null=True, help_text="Use this if you chose 'other'"
+        max_length=255,
+        blank=True,
+        null=True,
+        help_text="Use this if you chose 'other'",
+        validators=[validate_format_other],
```
Comment (tcouch) [inline mod_app/models/film_model.py:98]: I think this validator needs to be attached to format_type. Not sure if it'll be possible though this way. Maybe you need to override the clean method on the form?

File: museum_of_dreams_project/settings/aws.py [hunk 6]
```diff
@@ -23,3 +23,4 @@
         "PORT": os.environ["RDS_PORT"],
     }
 }
+GRAPPELLI_ADMIN_TITLE = "Museum of Dreamworlds"
```
Comment (tcouch) [inline museum_of_dreams_project/settings/aws.py:26]: why is this here and in base (and staging)?

**Review by tcouch — CHANGES_REQUESTED** [review]
Defo need to fix that migration issue.

File: mod_app/models/film_model.py (hunk 5 above)
Comment (acholyn) [reply mod_app/models/film_model.py:98]: I think I fixed this in a later PR - I attached it to the save method

File: mod_app/migrations/0001b_remove_link_models.py (hunk 3 above)
Comment (acholyn) [reply mod_app/migrations/0001b_remove_link_models.py:14]: Yes, this has been fixed in the bib PR; I didn't know about the issue until then because that's when I made some changes to the django workflow/action

File: mod_app/admin/film_admin.py (hunk 2 above)
Comment (acholyn) [reply mod_app/admin/film_admin.py:137]: Yes, the grey printed materials is acting as a header because you can't typically just choose where you want to put your inlines, I think custom templates would let me do that but I don't really want to go down that route unless absolutely necessary; Grappelli provides the next best thing

File: mod_app/admin/film_admin.py (hunk 1 above)
Comment (acholyn) [reply mod_app/admin/film_admin.py:54]: That might be a bit tricky since I'm overriding the field type and not defining each one individually to try keep it DRY

File: mod_app/admin/film_admin.py (hunk 2 above)
Comment (tcouch) [reply mod_app/admin/film_admin.py:137]: Yeah, fair enough.

File: mod_app/models/film_model.py (hunk 5 above)
Comment (tcouch) [reply mod_app/models/film_model.py:98]: I think a better option might be to attach a clean() method to the model:
https://docs.djangoproject.com/en/4.2/ref/models/instances/#validating-objects

That way the admin form should use it to validate the data and display a helpful error message.

File: museum_of_dreams_project/settings/aws.py (hunk 6 above)
Comment (tcouch) [reply museum_of_dreams_project/settings/aws.py:26]: Oh, I didn't notice the change from dreams to dreamworlds and back again.

File: mod_app/models/film_model.py (hunk 5 above)
Comment (acholyn) [reply mod_app/models/film_model.py:98]: have done on the bib branch 👍
